stacks-signer/src/v0/signer_state.rs

```diff
@@ -38,6 +40,205 @@ use crate::signerdb::SignerDb;
 /// This is the latest supported protocol version for this signer binary
 pub static SUPPORTED_SIGNER_PROTOCOL_VERSION: u64 = 0;
 
+/// A struct used to determine the current global state
+#[derive(Debug)]
+pub struct GlobalStateEvaluator {
+    /// A mapping of signer addresses to their corresponding vote weight
+    address_weights: HashMap<StacksAddress, u32>,
+    /// A mapping of signer addresses to their corresponding updates
+    address_updates: HashMap<StacksAddress, StateMachineUpdateMessage>,
+    /// The total weight of all signers
+    total_weight: u32,
+}
+
+impl GlobalStateEvaluator {
+    /// Create a new state evaluator
+    pub fn new(
+        address_updates: HashMap<StacksAddress, StateMachineUpdateMessage>,
+        address_weights: HashMap<StacksAddress, u32>,
+    ) -> Self {
+        let total_weight = address_weights
+            .values()
+            .fold(0u32, |acc, val| acc.saturating_add(*val));
+        Self {
+            address_weights,
+            address_updates,
+            total_weight,
+        }
+    }
+
+    /// Determine what the active signer protocol version should be
+    ///
+    /// NOTE: do not call this function unless the evaluator has already been updated with the current local state
+    pub fn determine_active_signer_protocol_version(&self) -> Option<u64> {
+        let mut protocol_versions = HashMap::new();
+        for (address, update) in &self.address_updates {
+            let Some(weight) = self.address_weights.get(address) else {
+                continue;
+            };
+            let entry = protocol_versions
+                .entry(update.local_supported_signer_protocol_version)
+                .or_insert_with(|| 0);
+            *entry += weight;
+            if *entry >= self.total_weight * 7 / 10 {
+                return Some(update.local_supported_signer_protocol_version);
+            }
+        }
+        None
+    }
+
+    /// Determine what the global burn view is if there is one
+    ///
+    /// NOTE: do not call this function unless the evaluator has already been updated with the current local state
+    pub fn determine_global_burn_view(&self) -> Option<(ConsensusHash, u64)> {
+        let mut burn_blocks = HashMap::new();
+        for (address, update) in &self.address_updates {
+            let Some(weight) = self.address_weights.get(address) else {
+                continue;
+            };
+            let StateMachineUpdateContent::V0 {
+                burn_block,
+                burn_block_height,
+                ..
+            } = update.content;
+
+            let entry = burn_blocks.entry(burn_block).or_insert_with(|| 0);
+            *entry += weight;
+            if *entry >= self.total_weight * 7 / 10 {
```

Comment:
Even though its a one-liner, I think it'd be good to add a function like:

```rust
/// Did the supplied vote weight cross the global agreement threshold?
pub fn reached_agreement(&self, vote_weight: u32) -> bool {
    vote_weight >= self.total_weight * 7 / 10
}
```

That way we're not duplicating the same arithmetic.

Reply:
Here i am reusing the same function in NakamotoBlockHeader: https://github.com/stacks-network/stacks-core/pull/5961/files#diff-b88f3619fd57d4dffe59b995552d694705c950aff495c7e9db7424c526a606f5R1522